src/main/scala/format/pgn/Parser.scala

```diff
@@ -98,296 +63,200 @@ object Parser {
+          valid(str)
+        case Left(err) => invalid(showExpectations("Cannot parse move", str, err))
       }
 
-    def strMove: Parser[StrMove] =
-      as("move") {
-        ((number | commentary) *) ~ forbidNullMove ~>
-          (moveRegex ~ nagGlyphs ~ rep(commentary) ~ nagGlyphs ~ rep(variation)) <~
-          (moveExtras *) ^^ { case san ~ glyphs ~ comments ~ glyphs2 ~ variations =>
-            StrMove(san, glyphs merge glyphs2, cleanComments(comments), variations)
-          }
-      }
+    val blockCommentary: P[String] = P.until0(P.char('}')).with1.between(P.char('{'), P.char('}'))
 
-    def number: Parser[String] = """[1-9]\d*[\s\.]*""".r
+    val inlineCommentary: P[String] = P.char(';') *> P.until(R.lf)
 
-    def moveExtras: Parser[Unit] =
-      as("moveExtras") {
-        commentary.^^^(())
-      }
+    val commentary = (blockCommentary | inlineCommentary).withContext("Invalid comment") <* whitespaces
 
-    def nagGlyphs: Parser[Glyphs] =
-      as("nagGlyphs") {
-        rep(nag) ^^ { nags =>
-          Glyphs fromList nags.flatMap { Glyph.find _ }
-        }
-      }
+    val result: P[String] = P.stringIn(List("*", "1/2-1/2", "½-½", "0-1", "1-0"))
 
-    val nagGlyphsRE = Glyph.PositionAssessment.all
-      .map(_.symbol)
-      .sortBy(-_.length)
-      .map(Regex.quote(_))
-      .mkString("|")
-      .r
+    val nagGlyphsRE = P.stringIn(
+      Glyph.PositionAssessment.all
+        .map(_.symbol)
+        .sortBy(-_.length)
+    )
 
-    def nag: Parser[String] =
-      as("nag") {
-        """\$\d+""".r | nagGlyphsRE
-      }
+    val nag = (P.char('$') ~ R.digit.rep).string | nagGlyphsRE
 
-    def variation: Parser[List[StrMove]] =
-      as("variation") {
-        "(" ~> strMoves <~ ")" ^^ { case (_, sms, _) => sms }
+    val nagGlyphs: P0[Glyphs] = (nag <* whitespaces).rep0.map(nags =>
+      Glyphs fromList nags.flatMap {
+        Glyph.find
       }
+    )
+
+    val moveExtras = commentary.as(())
+
+    val positiveIntString: P[String] =
+      (N.nonZeroDigit ~ N.digits0).string
+
+    // '. ' or '... ' or '. ... '
+    val numberSuffix = (P.char('.') | whitespace).rep0.void
+
+    // 10. or 10... but not 0 or 1-0 or 1/2
+    val number = (positiveIntString <* !P.charIn('-', '/') ~ numberSuffix).string
+
+    val forbidNullMove =
+      P.stringIn(List("--", "Z0", "null", "pass", "@@@@"))
+        .?
+        .flatMap(o => o.fold(P.unit)(_ => P.failWith("Lichess does not support null moves").void))
 
-    def commentary: Parser[String] = blockCommentary | inlineCommentary
+    val strMove: P[San] = P
+      .recursive[San] { recuse =>
+        val variation: P[Sans] =
+          (((P.char('(') <* whitespaces) *> recuse.rep0 <* (P.char(')') ~ whitespaces)) <* whitespaces)
+            .map(Sans(_))
 
-    def blockCommentary: Parser[String] =
-      as("block comment") {
-        "{" ~> """[^\}]*""".r <~ "}"
+        ((number.backtrack | commentary <* whitespaces).rep0 ~ forbidNullMove).with1 *>
```

Comment:
nit: I would put parens inside since it isn't really clear if `|` binds tighter than `<*`. I would argue it should not since `|` is more like addition and `<*` is a kind of product, so that should bind more tightly, but scala precedence is weird.

Reply:
`number` already has optional whitespaces so `<*` has higher priority is correct with my implementation. But I agree that add parens make this easier to understand.